# One-Hot Encoding 

### Import Packages

In [6]:
import pandas as pd
import numpy as np
import os 

### Load the Data Set

In [7]:
filename = os.path.join(os.getcwd(), "..", "..", "data", "cell2cell.csv")
df = pd.read_csv(filename, header=0)

#### Inspect the Data

In [8]:
df.shape

(51047, 58)

In [9]:
df.head()

CustomerID  Churn ServiceArea  ChildrenInHH  HandsetRefurbished  \
0     3000002   True   SEAPOR503         False               False   
1     3000010   True   PITHOM412          True               False   
2     3000014  False   MILMIL414          True               False   
3     3000022  False   PITHOM412         False               False   
4     3000026   True   OKCTUL918         False               False   

   HandsetWebCapable  TruckOwner  RVOwner  HomeownershipKnown  \
0               True       False    False                True   
1              False       False    False                True   
2              False       False    False               False   
3               True       False    False                True   
4              False       False    False                True   

   BuysViaMailOrder  ...  HandsetModels  CurrentEquipmentDays    AgeHH1  \
0              True  ...       0.487071             -0.077013  1.387766   
1              True  ...      -0.616775              3.019920  0.392039   
2             False  ...      -0.616775              3.019920 -0.241605   
3              True  ...       2.694763              0.305179 -0.060564   
4              True  ...       1.590917              1.857585  0.663601   

     AgeHH2  RetentionCalls  RetentionOffersAccepted  \
0 -0.883541        4.662897                  -0.1283   
1  0.871495       -0.180167                  -0.1283   
2  0.202910       -0.180167                  -0.1283   
3 -0.883541       -0.180167                  -0.1283   
4  1.372934       -0.180167                  -0.1283   

   ReferralsMadeBySubscriber  IncomeGroup  AdjustmentsToCreditRating  \
0                  -0.169283    -0.103411                  -0.140707   
1                  -0.169283     0.215243                  -0.140707   
2                  -0.169283     0.533896                  -0.140707   
3                  -0.169283     0.533896                  -0.140707   
4                  -0.169283     1.489856                   2.469282   

  HandsetPrice  
0    -0.864858  
1    -0.864858  
2    -0.368174  
3    -1.195980  
4    -1.195980  

[5 rows x 58 columns]

Our label will be the column `Churn`.

### Find the Columns (Features) That Contain String Values

In [10]:
df.dtypes

CustomerID                     int64
Churn                           bool
ServiceArea                   object
ChildrenInHH                    bool
HandsetRefurbished              bool
HandsetWebCapable               bool
TruckOwner                      bool
RVOwner                         bool
HomeownershipKnown              bool
BuysViaMailOrder                bool
RespondsToMailOffers            bool
OptOutMailings                  bool
NonUSTravel                     bool
OwnsComputer                    bool
HasCreditCard                   bool
NewCellphoneUser                bool
NotNewCellphoneUser             bool
OwnsMotorcycle                  bool
MadeCallToRetentionTeam         bool
CreditRating                  object
PrizmCode                     object
Occupation                    object
Married                       object
MonthlyRevenue               float64
MonthlyMinutes               float64
TotalRecurringCharge         float64
DirectorAssistedCalls        float64
O

The code cell below finds all columns of type `object`, which corresponds to the string type.

In [11]:
to_encode = list(df.select_dtypes(include=['object']).columns)
to_encode

['ServiceArea', 'CreditRating', 'PrizmCode', 'Occupation', 'Married']

## One-Hot Encode the Data Using NumPy

There are five object-type columns in our DataFrame `df`. Lets inspect the possible number of values each column (feature) may have.

In [12]:
df[to_encode].nunique()

ServiceArea     747
CreditRating      7
PrizmCode         4
Occupation        8
Married           2
dtype: int64

Notice that column `ServiceArea` has 747 potential values. This means we would have to create 747 new binary indicator columns - one column per unique value. That is too many!

Let's handle the special case of column `ServiceArea` first. Transforming this many categorical values would slow down the computation down the line. Instead, we will convert the top 10 most frequent values in column `ServiceArea`. 

In [13]:
top_10_SA = list(df['ServiceArea'].value_counts().head(10).index)

top_10_SA

['NYCBRO917',
 'HOUHOU281',
 'DALDAL214',
 'NYCMAN917',
 'APCFCH703',
 'DALFTW817',
 'SANSAN210',
 'APCSIL301',
 'SANAUS512',
 'SFROAK510']

Now that we have obtained the ten most frequent values for `ServiceArea`, let's transform DataFrame `df` to represent these values numerically.

The code cell below accomplishes the task of creating ten one-hot encoding columns. 

In [14]:
for value in top_10_SA:
    
    ## Create columns and their values
    df['ServiceArea_'+ value] = np.where(df['ServiceArea']==value,1,0)
    
    
# Remove the original column from DataFrame df
df.drop(columns = 'ServiceArea', inplace=True)

Inspect DataFrame `df` and see the new columns and their values.

In [15]:
df.head()

CustomerID  Churn  ChildrenInHH  HandsetRefurbished  HandsetWebCapable  \
0     3000002   True         False               False               True   
1     3000010   True          True               False              False   
2     3000014  False          True               False              False   
3     3000022  False         False               False               True   
4     3000026   True         False               False              False   

   TruckOwner  RVOwner  HomeownershipKnown  BuysViaMailOrder  \
0       False    False                True              True   
1       False    False                True              True   
2       False    False               False             False   
3       False    False                True              True   
4       False    False                True              True   

   RespondsToMailOffers  ...  ServiceArea_NYCBRO917  ServiceArea_HOUHOU281  \
0                  True  ...                      0                      0   
1                  True  ...                      0                      0   
2                 False  ...                      0                      0   
3                  True  ...                      0                      0   
4                  True  ...                      0                      0   

   ServiceArea_DALDAL214  ServiceArea_NYCMAN917  ServiceArea_APCFCH703  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   ServiceArea_DALFTW817  ServiceArea_SANSAN210  ServiceArea_APCSIL301  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

  ServiceArea_SANAUS512 ServiceArea_SFROAK510  
0                     0                     0  
1                     0                     0  
2                     0                     0  
3                     0                     0  
4                     0                     0  

[5 rows x 67 columns]

In [16]:
df.columns

Index(['CustomerID', 'Churn', 'ChildrenInHH', 'HandsetRefurbished',
       'HandsetWebCapable', 'TruckOwner', 'RVOwner', 'HomeownershipKnown',
       'BuysViaMailOrder', 'RespondsToMailOffers', 'OptOutMailings',
       'NonUSTravel', 'OwnsComputer', 'HasCreditCard', 'NewCellphoneUser',
       'NotNewCellphoneUser', 'OwnsMotorcycle', 'MadeCallToRetentionTeam',
       'CreditRating', 'PrizmCode', 'Occupation', 'Married', 'MonthlyRevenue',
       'MonthlyMinutes', 'TotalRecurringCharge', 'DirectorAssistedCalls',
       'OverageMinutes', 'RoamingCalls', 'PercChangeMinutes',
       'PercChangeRevenues', 'DroppedCalls', 'BlockedCalls', 'UnansweredCalls',
       'CustomerCareCalls', 'ThreewayCalls', 'ReceivedCalls', 'OutboundCalls',
       'InboundCalls', 'PeakCallsInOut', 'OffPeakCallsInOut',
       'DroppedBlockedCalls', 'CallForwardingCalls', 'CallWaitingCalls',
       'MonthsInService', 'UniqueSubs', 'ActiveSubs', 'Handsets',
       'HandsetModels', 'CurrentEquipmentDays', 'AgeHH1', 'AgeH

Let's inspect column `ServiceArea_DALDAL214` in row 47. Remember, it should have a value of 1.

In [17]:
df.loc[47]['ServiceArea_DALDAL214']

1

## One-Hot Encode the Data Using Pandas

Now that we have successfully transformed the `ServiceArea` column, let us transform the rest of the data in the remaining columns.

Let's begin with the`Married` column. Let's inspect the values in `Married`. 

In [18]:
df['Married']

0        False
1         True
2         True
3        False
4         True
         ...  
51042     True
51043    False
51044    False
51045    False
51046      NaN
Name: Married, Length: 51047, dtype: object

Note that each entry in the `Married` column contains one of two values - True or False. Therefore, we will replace the `Married` column with two new columns (one column per value).

In the code cell below, we are specifying which column to encode (`Married` column) as well as the prefix for the names of the new columns (`Married_`). Note that `pd.get_dummies()` returns a new DataFrame with the new one-hot encoded values. 

In [19]:
df_Married = pd.get_dummies(df['Married'], prefix='Married_')
df_Married

Married__False  Married__True
0                True          False
1               False           True
2               False           True
3                True          False
4               False           True
...               ...            ...
51042           False           True
51043            True          False
51044            True          False
51045            True          False
51046           False          False

[51047 rows x 2 columns]

Since the `pd.get_dummies()` function returned a new DataFrame rather than making the changes to the original DataFrame `df`, let us add the new DataFrame `df_Married` to DataFrame `df`, and delete the original `Married` column from DataFrame `df`.


In [20]:
# Concatenate DataFrame df with the one-hot encoded DataFrame df_Married
df = df.join(df_Married)

# Remove the original 'Married' column from DataFrame df
df.drop(columns = 'Married', inplace=True)


Let's inspect DataFrame `df` to see the changes that have been made. DataFrame `df` now contains columns `Married__True` and `Married__False` and no longer contains the `Married` column.

In [21]:
df.columns

Index(['CustomerID', 'Churn', 'ChildrenInHH', 'HandsetRefurbished',
       'HandsetWebCapable', 'TruckOwner', 'RVOwner', 'HomeownershipKnown',
       'BuysViaMailOrder', 'RespondsToMailOffers', 'OptOutMailings',
       'NonUSTravel', 'OwnsComputer', 'HasCreditCard', 'NewCellphoneUser',
       'NotNewCellphoneUser', 'OwnsMotorcycle', 'MadeCallToRetentionTeam',
       'CreditRating', 'PrizmCode', 'Occupation', 'MonthlyRevenue',
       'MonthlyMinutes', 'TotalRecurringCharge', 'DirectorAssistedCalls',
       'OverageMinutes', 'RoamingCalls', 'PercChangeMinutes',
       'PercChangeRevenues', 'DroppedCalls', 'BlockedCalls', 'UnansweredCalls',
       'CustomerCareCalls', 'ThreewayCalls', 'ReceivedCalls', 'OutboundCalls',
       'InboundCalls', 'PeakCallsInOut', 'OffPeakCallsInOut',
       'DroppedBlockedCalls', 'CallForwardingCalls', 'CallWaitingCalls',
       'MonthsInService', 'UniqueSubs', 'ActiveSubs', 'Handsets',
       'HandsetModels', 'CurrentEquipmentDays', 'AgeHH1', 'AgeHH2',
      

Let's use the approach to one-hot encode the remain columns `CreditRating`, `PrizmCode` and `Occupation`.

#### `CreditRating` Column

We saw earlier that each row in the `CreditRating` column contains one of seven possible values. Therefore, we will replace the `CreditRating` column with seven new columns (one column per value).

In [22]:
# Inspect the values in the column
df['CreditRating']

0        1-Highest
1         4-Medium
2           3-Good
3         4-Medium
4        1-Highest
           ...    
51042    1-Highest
51043       3-Good
51044        5-Low
51045        5-Low
51046        5-Low
Name: CreditRating, Length: 51047, dtype: object

In [23]:
# Use pd.get_dummies() to create a new DataFrame with the one-hot encoded values.
df_CreditRating = pd.get_dummies(df['CreditRating'], prefix='CreditRating_')
df_CreditRating

CreditRating__1-Highest  CreditRating__2-High  CreditRating__3-Good  \
0                         True                 False                 False   
1                        False                 False                 False   
2                        False                 False                  True   
3                        False                 False                 False   
4                         True                 False                 False   
...                        ...                   ...                   ...   
51042                     True                 False                 False   
51043                    False                 False                  True   
51044                    False                 False                 False   
51045                    False                 False                 False   
51046                    False                 False                 False   

       CreditRating__4-Medium  CreditRating__5-Low  CreditRating__6-VeryLow  \
0                       False                False                    False   
1                        True                False                    False   
2                       False                False                    False   
3                        True                False                    False   
4                       False                False                    False   
...                       ...                  ...                      ...   
51042                   False                False                    False   
51043                   False                False                    False   
51044                   False                 True                    False   
51045                   False                 True                    False   
51046                   False                 True                    False   

       CreditRating__7-Lowest  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
...                       ...  
51042                   False  
51043                   False  
51044                   False  
51045                   False  
51046                   False  

[51047 rows x 7 columns]

In [24]:
# Concatenate DataFrame df with the one-hot encoded DataFrame df_CreditRating
df = df.join(df_CreditRating)

# Remove the original 'CreditRating' column from DataFrame df
df.drop(columns = 'CreditRating', inplace=True)


#### `PrizmCode` Column

Note that each row in the `PrizmCode` column contains one of four possible values. Therefore, we will replace the `PrizmCode` column with four new columns (one column per value).

In [25]:
# Inspect the values in the column
df['PrizmCode']

0        Suburban
1        Suburban
2            Town
3           Other
4           Other
           ...   
51042    Suburban
51043       Other
51044       Other
51045       Other
51046       Other
Name: PrizmCode, Length: 51047, dtype: object

In [26]:
# Use pd.get_dummies() to create a new DataFrame with the one-hot encoded values.
df_PrizmCode = pd.get_dummies(df['PrizmCode'], prefix='PrizmCode_')
df_PrizmCode

PrizmCode__Other  PrizmCode__Rural  PrizmCode__Suburban  \
0                 False             False                 True   
1                 False             False                 True   
2                 False             False                False   
3                  True             False                False   
4                  True             False                False   
...                 ...               ...                  ...   
51042             False             False                 True   
51043              True             False                False   
51044              True             False                False   
51045              True             False                False   
51046              True             False                False   

       PrizmCode__Town  
0                False  
1                False  
2                 True  
3                False  
4                False  
...                ...  
51042            False  
51043            False  
51044            False  
51045            False  
51046            False  

[51047 rows x 4 columns]

In [27]:
# Concatenate DataFrame df with the one-hot encoded DataFrame df_PrizmCode
df = df.join(df_PrizmCode)

# Remove the original 'PrizmCode' column from DataFrame df
df.drop(columns = 'PrizmCode', inplace=True)


#### `Occupation` Column

Note that each row in the `Occupation` column contains one of eight possible values. Therefore, we will replace the `Occupation` column with eight new columns (one column per value).


In [28]:
# Inspect the values in the column
df['Occupation']

0        Professional
1        Professional
2              Crafts
3               Other
4        Professional
             ...     
51042           Other
51043           Other
51044        Clerical
51045           Other
51046           Other
Name: Occupation, Length: 51047, dtype: object

In [29]:
# Use pd.get_dummies() to create a new DataFrame with the one-hot encoded values.
df_Occupation = pd.get_dummies(df['Occupation'], prefix='Occupation_')
df_Occupation

Occupation__Clerical  Occupation__Crafts  Occupation__Homemaker  \
0                     False               False                  False   
1                     False               False                  False   
2                     False                True                  False   
3                     False               False                  False   
4                     False               False                  False   
...                     ...                 ...                    ...   
51042                 False               False                  False   
51043                 False               False                  False   
51044                  True               False                  False   
51045                 False               False                  False   
51046                 False               False                  False   

       Occupation__Other  Occupation__Professional  Occupation__Retired  \
0                  False                      True                False   
1                  False                      True                False   
2                  False                     False                False   
3                   True                     False                False   
4                  False                      True                False   
...                  ...                       ...                  ...   
51042               True                     False                False   
51043               True                     False                False   
51044              False                     False                False   
51045               True                     False                False   
51046               True                     False                False   

       Occupation__Self  Occupation__Student  
0                 False                False  
1                 False                False  
2                 False                False  
3                 False                False  
4                 False                False  
...                 ...                  ...  
51042             False                False  
51043             False                False  
51044             False                False  
51045             False                False  
51046             False                False  

[51047 rows x 8 columns]

In [30]:
# Concatenate DataFrame df with the one-hot encoded DataFrame df_Occupation
df = df.join(df_Occupation)

# Remove the original 'Occupation' column from DataFrame df
df.drop(columns = 'Occupation', inplace=True)


Let's now inspect DataFrame `df` and see how the columns have been transformed. Notice the new dimensions of `df`: we previously had 58 columns and now we have 84.

In [31]:
df.columns

Index(['CustomerID', 'Churn', 'ChildrenInHH', 'HandsetRefurbished',
       'HandsetWebCapable', 'TruckOwner', 'RVOwner', 'HomeownershipKnown',
       'BuysViaMailOrder', 'RespondsToMailOffers', 'OptOutMailings',
       'NonUSTravel', 'OwnsComputer', 'HasCreditCard', 'NewCellphoneUser',
       'NotNewCellphoneUser', 'OwnsMotorcycle', 'MadeCallToRetentionTeam',
       'MonthlyRevenue', 'MonthlyMinutes', 'TotalRecurringCharge',
       'DirectorAssistedCalls', 'OverageMinutes', 'RoamingCalls',
       'PercChangeMinutes', 'PercChangeRevenues', 'DroppedCalls',
       'BlockedCalls', 'UnansweredCalls', 'CustomerCareCalls', 'ThreewayCalls',
       'ReceivedCalls', 'OutboundCalls', 'InboundCalls', 'PeakCallsInOut',
       'OffPeakCallsInOut', 'DroppedBlockedCalls', 'CallForwardingCalls',
       'CallWaitingCalls', 'MonthsInService', 'UniqueSubs', 'ActiveSubs',
       'Handsets', 'HandsetModels', 'CurrentEquipmentDays', 'AgeHH1', 'AgeHH2',
       'RetentionCalls', 'RetentionOffersAccepted',
    

In [32]:
df.shape

(51047, 84)